src/User/UserProvider.php

```diff
@@ -25,7 +25,7 @@ public function __construct(
     /**
      * {@inheritdoc}
      */
-    public function loadUserByUsername($username): UserInterface
+    public function loadUserByUsername($username): ?UserInterface
```

Comment:
Thanks for the PR maybe its useful to use less null values and throw an exception here when the user was not found.

Reply:
If you look in line 51, it is okay to not find a user and to return null

Reply:
I know that the repository function in general is returning `object` or `null`. This is a design decision based this article. https://afilina.com/learn/nulls/return-types It is a really good description of Fail Fast and Use Exceptions.

i could imagine something like:
```php
<?php
// ...
public function loadUserByUsername($username): UserInterface
{
     /** UserInterface|null $user */
     $user = $this->userManager->findUserByUsernameOrEmail($username);

     if (null === $user) {
          throw UserNotFound::createWithUsername($username)
     }

     return $user;
}

```

Reply:
And in this case you have even more strict types. and you can prevent the null check everywhere else by catching the exception.

Reply:
Sounds good to me

Reply:
You want to do that?